keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,806 @@
+In a nutshell, RootlessKit is an extended version of [`unshare -rmn`](http://man7.org/linux/man-pages/man1/unshare.1.html) for rootless containers.
+RootlessKit has been already adopted by Docker, BuildKit, Usernetes, k3s, and partially by Podman.
+
+```console
+$ rootlesskit --net=slirp4netns --copy-up=/etc --copy-up=/run --copy-up=/var ... sleep infinity &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -U -n -m kubelet ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -U -n -m containerd ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -U -n -m kube-proxy ... &
+$ nsenter -t $ROOTLESSKIT_CHILD_PID -U -n -m flanneld ... &
+```
+
+> **Note**
+>
+> Rootless mode is unrelated to [the Node-level UserNS KEP](https://github.com/kubernetes/enhancements/issues/127), and they do not mutually conflict.
+> See [Non-goals](#non-goals) section of this KEP.
+
+#### Paths
+
+Some paths like `/var/log/pods` are hardcoded in Kubernetes and hard to change.
+
+Although these directories are not writable by unprivileged users, Kubernetes does NOT need to be changed to use unprivileged home directories,
+because RootlessKit can bind-mount writable directories on these paths without the root privileges. (`rootlesskit --copy-up=/var`)
+
+#### Network
+
+The node components need to be executed in a network namespace created along with the user namespace, because an unprivileged user cannot do privileged operations in the host network namespace.
+As the components are executed inside a network namespace, `NodePorts` are not directly accessible from other hosts.
+
+An external controller should watch changes on `corev1.Service` resources and call [RootlessKit API](https://github.com/rootless-containers/rootlesskit/blob/v0.9.5/pkg/api/openapi.yaml) to set up port forwarding for the node ports.
+
+k3s implementation: https://github.com/rancher/k3s/blob/v1.17.2+k3s1/pkg/rootlessports/controller.go#L92-L96
+
+##### RootlessKit network drivers
+
+RootlessKit supports two kinds of network stacks:
+* TAP with pure usermode network stack (either `slirp4netns` or VPNKit)
+* vEth with setuid binary `lxc-user-nic`
+
+`slirp4netns` is preferred for security, `lxc-user-nic` is preferred for performance.
+
+These stacks are used for the namespace where the node components are executed in, not for the containers' namespaces.
+CNI plugins such as Flannel are expected to be used for the containers' namespace.
+
+##### CNI plugins
+
+Flannel (VXLAN) is known to work.
+
+#### Pod SecurityContext
+
+- `runAsUser`: supported, but the number of the UID is limited by `/etc/subuid`.
+- `sysctls`: some sysctl parameters are supported, but some would fail in `EPERM`.
+  Creating Pod manifests with such sysctl parameters would fail.
+  If this behavior is problematic, user should write a Mutating Admission Webhook to remove such sysctl parameters from Pod manifests.
+- seccomp: supported
+- AppArmor: unsupported. Creating Pod manifests with AppArmor profile specified would fail.
```

Comment:
any input around selinux?

Reply:
Same status as Rootless Podman. i.e. Applying an existing profile would be ok, but creating a new profile would not.